sql/sql_admission_control.h

```diff
@@ -0,0 +1,356 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#ifndef _sql_admission_control_h
+#define _sql_admission_control_h
+
+#include "mysql/psi/mysql_mutex.h"
+#include "mysql/psi/psi_base.h"
+
+#include <deque>
+#include "sql/sql_class.h"
+
+#define IS_BIT_SET(val, n) ((val) & (1 << (n)))
+
+extern bool opt_admission_control_by_trx;
+extern ulonglong admission_control_filter;
+class AC;
+extern AC *db_ac;
+
+#ifdef HAVE_PSI_INTERFACE
+extern PSI_stage_info stage_admission_control_enter;
+extern PSI_stage_info stage_admission_control_exit;
+extern PSI_stage_info stage_waiting_for_admission;
+#endif
+
+// TODO: remove this block eventually when this becomes part of audit plugin?
+// TBD later
+// Resource isolation types that a multi-tenancy plugin will handle.
+// Currently connection and query limits are two resource types. More will be
+// supported in the future.
+enum class enum_multi_tenancy_resource_type : int32_t {
+  MULTI_TENANCY_RESOURCE_CONNECTION,
+  MULTI_TENANCY_RESOURCE_QUERY,
+
+  MULTI_TENANCY_NUM_RESOURCE_TYPES
+};
+
+// Callback function return types.
+// - ACCEPT: resource can be granted
+// - WAIT: may need to wait for resource to be freed up
+// - REJECT: resource cannot be granted
+// - FALLBACK: plugin is disabled
+enum class enum_multi_tenancy_return_type : int32_t {
+  MULTI_TENANCY_RET_ACCEPT = 0,
+  MULTI_TENANCY_RET_WAIT,
+  MULTI_TENANCY_RET_REJECT,
+
+  MULTI_TENANCY_RET_FALLBACK,
+  MULTI_TENANCY_NUM_RETURN_TYPES
+};
+
+typedef enum enum_multi_tenancy_resource_type MT_RESOURCE_TYPE;
+typedef enum enum_multi_tenancy_return_type MT_RETURN_TYPE;
+typedef std::unordered_map<std::string, std::string> ATTRS_MAP_T;
+typedef std::vector<std::pair<std::string, std::string>> ATTRS_VEC_T;
+
+struct multi_tenancy_resource_attributes {
+  const ATTRS_MAP_T *connection_attrs_map;
+  const ATTRS_VEC_T *query_attrs_vec;
+  const char *database;
+};
+
+typedef struct multi_tenancy_resource_attributes MT_RESOURCE_ATTRS;
+
+// The contents here must match entries in admission_control_filter_names array
+enum enum_admission_control_filter {
+  ADMISSION_CONTROL_ALTER,
+  ADMISSION_CONTROL_BEGIN,
+  ADMISSION_CONTROL_COMMIT,
+  ADMISSION_CONTROL_CREATE,
+  ADMISSION_CONTROL_DELETE,
+  ADMISSION_CONTROL_DROP,
+  ADMISSION_CONTROL_INSERT,
+  ADMISSION_CONTROL_LOAD,
+  ADMISSION_CONTROL_SELECT,
+  ADMISSION_CONTROL_SET,
+  ADMISSION_CONTROL_REPLACE,
+  ADMISSION_CONTROL_ROLLBACK,
+  ADMISSION_CONTROL_TRUNCATE,
+  ADMISSION_CONTROL_UPDATE,
+  ADMISSION_CONTROL_SHOW,
+  ADMISSION_CONTROL_USE,
+  ADMISSION_CONTROL_END = 64
+};
+
+/*
+ * sql_admission_control.h/cc
```

Comment:
Was this supposed to be a file header comment?

Reply:
No idea. It was there in original FB source. You want me to change this?